### PR TITLE
NamespacedId format

### PR DESCRIPTION
**Pull request by ltj**
Namespaced IDs is likely to be used as identifiers in the infrastructure. As such it makes sense to apply a stricter format that makes use of canonical strings less error-prone.

The suggestion here is to use a format similar to that of Java package names, and in lower case.

`ParticipantAttribute` used a UUID as the name. In order to support this usecase and comply with the stricter format, a `UUID.toNonDashedString()` function has been added to allow using UUIDs as "words" without hyphens.

### CODE REVIEW

**Comment by Whathecode** [comment]
Awesome! `NamespacedId` was indeed designed to be more restricted and we just didn't get around to implementing it yet.

I just [read this](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html):

> In some cases, the internet domain name may not be a valid package name. This can occur if the domain name contains a hyphen or other special character, if the package name begins with a digit or other character that is illegal to use as the beginning of a Java name, or if the package name contains a reserved Java keyword, such as "int". In this event, the suggested convention is to add an underscore.

I don't dislike the `toNonDashedString()` approach, but if the goal is to comply with the Java standard, replacing such symbols with underscores could provide some consistency/expected behavior to API users.

File: carp.common/src/commonTest/kotlin/dk/cachet/carp/common/application/NamespacedIdTest.kt [hunk 1]
```diff
@@ -26,6 +26,43 @@ class NamespacedIdTest
...
+            NamespacedId( "some.namespace", "contains-hyphen!" )
+        }
+    }
+
+    @Test
+    fun namespace_cannot_end_with_dot()
+    {
+        assertFailsWith<IllegalArgumentException>
+        {
+            NamespacedId("some.namespace.", "name")
```
Comment (Whathecode) [inline carp.common/src/commonTest/kotlin/dk/cachet/carp/common/application/NamespacedIdTest.kt:43]: Detekt might throw some formatting errors here (spaces should be needed). I can take care of those, don't worry. :)

**Review by Whathecode — COMMENTED** [review]
Overall looks good!

But with the Java documentation on replacing invalid characters with underscores I found, I am now contemplating whether we should provide similar functionality instead of `toNonDashString`. 🤔 What do you think?

Now that I know underscores are allowed, I'm also contemplating [adding underscores to some of the types in `CarpDataTypes`](https://github.com/cph-cachet/carp.core-kotlin/blob/develop/carp.common/src/commonMain/kotlin/dk/cachet/carp/common/application/data/CarpDataTypes.kt#L18), but that isn't part of this PR.

File: carp.common/src/commonTest/kotlin/dk/cachet/carp/common/application/NamespacedIdTest.kt [hunk 2]
```diff
@@ -26,6 +26,43 @@ class NamespacedIdTest
         }
     }
 
+    @Test
+    fun name_may_not_contain_non_alphanumerics()
+    {
+        assertFailsWith<IllegalArgumentException>
+        {
+            NamespacedId( "some.namespace", "contains-hyphen!" )
```
[the remark this answers is not in the document]
Comment (ltj) [reply carp.common/src/commonTest/kotlin/dk/cachet/carp/common/application/NamespacedIdTest.kt:34]: Deliberately. But yes :)

**Pull request merged**